platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/Bucket.java

```diff
@@ -246,242 +191,233 @@ public long findValue(final int keyHashCode, final K key, final long notFoundVal
                         + value
                         + " keyHashCode="
-                        + readHash
-                        + " keyVer="
-                        + keySerializationVersion
+                        + hashCode
                         + " key="
                         + key
-                        + " keySize="
-                        + keySize
                         + "\n");
-                entryOffset += KEY_HASHCODE_SIZE + VALUE_SIZE + keySize;
             }
         } catch (IOException e) {
             logger.error(EXCEPTION.getMarker(), "Failed enumerating bucket entries", e);
         }
-        bucketBuffer.clear();
-        sb.append("} RAW DATA = ");
-        final byte[] bucketArray = bucketBuffer.array();
-        for (int i = 0; i < size; i++) {
-            sb.append(String.format("%02X ", bucketArray[i]).toUpperCase());
-        }
+        sb.append("}");
         return sb.toString();
     }
 
-    // =================================================================================================================
-    // FindResult class
-
     /**
-     * Simple record for entry lookup results. If an entry is found, "found" is set to true,
-     * "entryOffset" is the entry offset in bytes in the bucket buffer, "entryIndex" is the
-     * entry index in the array of entries, and "entryValue" is the entry value. If no entity
-     * is found, "found" is false, "entryOffset" is the total size of the bucket buffer,
-     * "entryIndex" and "entryValue" are undefined.
+     * A single entry in a bucket, which contains key hash code, value (usually, path), and
+     * full serialized key. A bucket may contain multiple such entries.
+     *
+     * This class would be a record, if it was immutable. However, when a value is updated
+     * in a bucket, and a bucket entry already exists for the same key, instead of creating
+     * a new entry, we just update the value in the existing entry.
      */
-    private record FindResult(int entryOffset, int entryIndex, boolean found, long entryValue) {}
+    private class BucketEntry {
+
+        /** Key hash code */
+        private int hashCode;
+        /** Long value */
+        private long value;
+        /** Serialized key */
+        private BufferedData keyBytes;
+
+        /** Creates new bucket entry from hash code, value, and serialized key bytes */
+        public BucketEntry(final int hashCode, final long value, @NonNull final BufferedData keyBytes) {
+            this.hashCode = hashCode;
+            this.value = value;
+            this.keyBytes = Objects.requireNonNull(keyBytes);
+        }
+
+        /** Creates new bucket entry by reading its fields from the given buffer */
+        public BucketEntry(final BufferedData entryData) {
+            // defaults
+            hashCode = 0;
+            value = 0;
+            keyBytes = null;
+
+            // read fields
+            while (entryData.hasRemaining()) {
```

Comment:
Just to confirm my understanding - entry data is supposed to have one or two tags depending on the type. In case of the hash code it's one, in case of leaf entry it's two - value and key. Is that right?

Reply:
Buckets contain entries. Each entry contains a key (bytes), a key hash code, and a value (path). It's object key to path mapping, after all. When an entry is deserialized, these three fields may occur in any order in the source buffer. One field, value (path), is optional and may not be present at all.

This loop solved these issues. First, a tag is read. Then a field number is extracted from the tag. Then based on the field number, the right field is read. This pattern is common across all classes that deserialize data from PBJ buffers, e.g. `VirtualLeafSerializer`